src/mds/MDSRank.cc

```diff
@@ -325,11 +325,15 @@ class C_Drop_Cache : public MDSInternalContext {
              << " recalled " << count << " caps" << dendl;
 
     caps_recalled += count;
+    uint64_t remaining = (recall_timeout == 0 ? 0 : recall_timeout-duration);
     if ((throttled || count > 0) && (recall_timeout == 0 || duration < recall_timeout)) {
-      auto timer = new FunctionContext([this](int _) {
-        recall_client_state();
-      });
-      mds->timer.add_event_after(1.0, timer);
+      C_ContextTimeout *ctx = new C_ContextTimeout(
+        mds, std::min<uint64_t>(1, remaining), new FunctionContext([this](int r) {
```

Comment:
when `remaining` is zero, `start_timer()` does not schedule the timer task, thus never completing the context.

Reply:
I don't think we can hit that condition because of the `duration < recall_timeout` check above. In any case, allowing remaining to underflow is a code smell. I'll fix that.

Reply:
fixed